### PR TITLE
RTCRtpEncodingParameters attribute to turn on/off sending CN / DTX

### PR DESCRIPTION
**Pull request by aboba**
Fix for Issue https://github.com/w3c/webrtc-pc/issues/644

### CODE REVIEW

**Comment by fluffy** [comment]
Like this but think it might need a few more things. We need to say how it interacts wth policy set, I think it needs to say it just overrides any policy. It also needs to be clear that if the SDP did not negotiate the ability to send CN/ DTX, then it will not be sent regardless of how this is set.

**Comment by pthatcherg** [comment]
Would a name of ".comfortNoise" make sense with an enum value of active/inactive make sense?

I agree it should go on the encoding parameters.

**Comment by stefhak** [comment]
I think we need to sort out what should happen if CN/DTX has *not* been negotiated, and this flag is set to true, or if CN/DTX has been negotiated but is negotiated away later (with this flag true). Will we report errors?

**Comment by pthatcherg** [comment]
Well, perhaps that's why an enum might be good.  Then we could have "disable" and "enableIfNegotiated" or something like that.

**Comment by stefhak** [comment]
@pthatcherg I agree, that would simplify.

**Comment by aboba** [comment]
@pthatcherg Like the enum idea.

**Comment by aboba** [comment]
@pthatcherg In ORTC, DTX is controlled by setting a codec parameter (e.g. in the case of Opus, parameters.useDtx).  Since CN usage can be controlled by adding/removing the CN codec via  setParameters(), should we consider a similar parameter for Opus in WebRTC 1.0, rather than an encoding parameter?

**Comment by aboba** [comment]
@stefhak @dontcallmedom Are there any good tutorials on the new respec webidl formatting?  Clearly I do not understand this well enough.

File: webrtc.html [hunk 1]
```diff
@@ -4830,11 +4842,39 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
         </section>
       </div>
       <div>
+        <pre class="idl">enum RTCVadStatus {
+        "disable",
+        "enableIfNegotiated"
+        };</pre>
```
Comment (dontcallmedom) [inline webrtc.html]: you use "disable" and "enable" here, but "disabled" and "enabled" in the table below (the latter sounds better)

**Comment by dontcallmedom** [comment]
The reference documentation on the new WebIDL is at https://www.w3.org/respec/examples/webidl-contiguous.html

Note that you're using it right, the problem here is a consistency between the IDL declaration and the names used in the description.

**Comment by fluffy** [comment]
the PR  LGTM but wanted to comment on name comfortNoise ... there are actually two things going on here. 1) is when no one is speaking, do you send the background noise or suppress it out and send nothing - that is Voice Activity Detection. 2) the other things is when the receiver receives the evidence of suppressed packets - so it effectively it has received a digital zero - does it it generate a synthetic background noise to insert into the audio it is playing. That is comfort noise.  When phones first switched to digital signals, the background noise was so low compared to analog lines that people thought the call had disconnected to "comfort noise" was added to the call to simulate the noise on analog phone lines. For high def audio, applications are slowly migrating away from comfort noise but continuing to do voice activity detection. I think the thing we want to control here is voice activity detection not comfort noise.

**Comment by pthatcherg** [comment]
@fluffy How about .discontinuousTransmission instead of .comfortNoise?  That more closely matches what opus calls it, and it covers all the use cases you are describing.  I'd favor that over "voiceActivityDetection", because technically it doesn't have to be "voice" that activates/deactivates it.  It could be instrumental music or something else as well.  

@aboba There are two problems with using codec parameters in WebRTC 1.0:  1.  So far we have said codecs can't be changed, only reordered or removed, so we'd have to deal with that.  2.  It's awkward to say "for codec X, frob A, for codec Y, frob B".  It would be nice to have one frob for turning this on and off.

**Comment by pthatcherg** [comment]
I thought about this a little more based on Cullen's comment.  Would it make sense to have 3 enum values and rename it this?

enum RTCTransmissionContinuity {
  "continuous",
  "discontinuous"
  // Discontinuous if CN not possible
  "discontinuousWithComfortNoise"
};

dictionary RTCRtpEncodingParameters {
  ...
  RTCTransmissionContinuity transmissionContinuity = "continuous";
}

**Comment by aboba** [comment]
@pthatcherg Like the idea of consistency between codecs.  I understand settings to turn on DTX or turn off DTX. However I am unclear about the interaction with existing SDP negotiated CN. For some codecs (like Opus) couldn't you turn on DTX even if CN was not negotiated? How does discontinuousWithComfortNoise work if Opus is negotiated?

**Comment by stefhak** [comment]
@aboba and @pthatcherg reading https://tools.ietf.org/html/draft-ietf-payload-rtp-opus-11 I get the impression that CN is not used with Opus, but comfort noise is instead generated by the decoder if no speech frames are received (like if DTX is enabled and no speech signal is present).

**Comment by aboba** [comment]
@fluffy "if the SDP did not negotiate the ability to send CN/ DTX, then it will not be sent"
Is this still true in the case of Opus? Or would the attribute still affect use of DTX in that case even if CN were not negotiated?

**Comment by stefhak** [comment]
Looking at codecs like AMR, AMR-WB and EVS it is unclear on if, and how, this API should be mapped. These codecs, to the best of my understanding, roughly work something like:

- DTX (Discontinuous Transmission) is built in

- for EVS it is possible to switch DTX off, but for AMR/AMR-WB default implementations run with DTX all the time

- whenever the VAD detects "no speech" DTX kicks in, and during DTX comfort noise data is sent by so called SID (SIlence Descriptor) frames. (There are two types, SID-first and SID-update but that does not matter that much.)

- when entering DTX a SID frame is generated followed by another one almost immediately after. From that point a new SID frame is generated every 8th frame (160ms)

- the decoder end uses the SID frames to generate comfort noise

- in absence of SID frames comfort noise would be generated by the decoder using the last speech frames received (I think this is similar to Opus) - there is a hangover period used meaning that the last few speech frames before DTX usually have no speech in them - for some time, then the decoder would mute its output signal.

- Turning DTX off would mean that speech frames (carrying the background noise) would be transmitted.

Assuming the use case is to have the decoder generate silence instead of comfort noise when there is no speech input signal this means: 

* Turning DTX off should be straightforward for EVS, and could, depending on implementation support in the codec lib, be possible for AMR/AMR-WB as well (but is not by default).

** *But* even if DTX is turned off the background noise would any way be carried to the receiver in speech frames.

* Suppressing the sending of the SID frames could be possible (e.g. by not forwarding them to the RTP stack), but then comfort noise would anyway be generated at the decoder side for some time (160ms - 1s are figures I have heard) before it mutes.

* The only way that seems possible is to replace SID frames for codec homing frames or modify the SID frames such that the decoder produce silence. However, that feels like a hack to me.

**Comment by aboba** [comment]
@stefhak Thank you for the research. 

Does it make sense to:

a. Have a control to turn DTX on/off on an RtpSender.  On a codec with internal DTX this would turn off the internal DTX mechanism and on a codec without internal DTX it would turn off the sending of CN (and therefore could only be enabled if CN was negotiated).  

b. Have a control to turn comfort noise decoding on/off on an RtpReceiver. On a codec with internal DTX, this would substitute silence for internal comfort noise in the decoder and on a codec without internal DTX it would turn off the decoding of CN (if CN was negotiated).

**Comment by stefhak** [comment]
@aboba I think that proposal makes sense.

One question is if you need a) if you have b) (assuming default is to use DTX if possible).

I worry a bit about error handling, would this knob exist on a video RtpSender, and what would happen if it was turned?

**Comment by aboba** [comment]
@stefhak   a) is about the encoder, b) is about the decoder.  If internal DTX or CN is not being generated by the encoder, then the decoder will not decode CN or internal DX, no matter what the setting is. However, if internal DTX or CN is being generated by the encoder, then the decoder could choose not to decode it and substitute silence instead.  The question is whether this is useful or not.

These controls do not make sense on a video RtpSender/Receiver so the values would presumably be ignored.

File: webrtc.html [hunk 2]
```diff
@@ -4741,6 +4741,7 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
              unsigned long       ssrc;
              RTCRtpRtxParameters rtx;
              RTCRtpFecParameters fec;
+             RTCDtxStatus        useDtx;
```
Comment (pthatcherg) [inline webrtc.html]: Instead of "useDtx", can we just have "dtx"?

**Comment by aboba** [comment]
@juberti @pthatcherg  Is it ok now?

File: webrtc.html [hunk 3]
```diff
@@ -4830,11 +4844,39 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
         </section>
       </div>
       <div>
+        <pre class="idl">enum RTCDtxStatus {
+        "disabled",
+        "enabledIfNegotiated"
```
Comment (pthatcherg) [inline webrtc.html]: I'm fine with this, but I thought Justin's suggestion was to just make this "enabled" so it was less verbose.  I'm fine either way.

**Comment by pthatcherg** [comment]
LGTM

**Comment by aboba** [comment]
@pthatcherg  Took Justin's suggestion.  Changed enum to "enabled" with the same description (only enabled if negotiated).

**Pull request closed without merging**